File: benchmarks/cross-language/go/skynet.go
// Go Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Recursive tree of goroutines: root spawns 10 children, each spawns 10 more, etc.
// Leaves report their offset; parents aggregate 10 children's results and report up.
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func getLeaves() int64 {
	if env := os.Getenv("SKYNET_LEAVES"); env != "" {
		if n, err := strconv.ParseInt(env, 10, 64); err == nil {
			return n
		}
	}
	if env := os.Getenv("BENCHMARK_MESSAGES"); env != "" {
		if n, err := strconv.ParseInt(env, 10, 64); err == nil {
			return n
		}
	}
	return 1000000
}

// skynetNode sends its subtree sum to the result channel.
// Leaves send their offset directly; internal nodes spawn 10 children and sum.
// seqSum sums a subtree on the current goroutine, no further spawning.
const seqThreshold int64 = 1000

func seqSum(offset, size int64) int64 {
	if size == 1 {
		return offset
	}
	childSize := size / 10
	var sum int64
	for i := int64(0); i < 10; i++ {
		sum += seqSum(offset+i*childSize, childSize)
	}
	return sum
}

func skynetNode(result chan<- int64, offset, size int64) {
	// Same threshold as every other language in this suite. Go could spawn all
	// 1,111,111 goroutines and used to, but then it did a thousand times the
	// concurrency work of the pthread implementations while being scored on the
	// same divisor, which is why it came last on a benchmark it should win.
	if size <= seqThreshold {
		result <- seqSum(offset, size)
		return
	}
	children := make(chan int64, 10)
	childSize := size / 10
	for i := int64(0); i < 10; i++ {
		go skynetNode(children, offset+i*childSize, childSize)
	}
	var sum int64
	for i := 0; i < 10; i++ {
		sum += <-children
	}
	result <- sum
}

func main() {
	numLeaves := getLeaves()

	// Concurrency units actually created; reported, never used as a divisor.
	totalActors := int64(1)
	for n := numLeaves; n > seqThreshold; n /= 10 {
		totalActors += n / seqThreshold
	}
	// Divide by the units created, not the tree's node count. Every language in
	// this suite now uses the same threshold, so the counts are equal and a
	// per-unit cost is directly comparable.
	rateBase := totalActors

	fmt.Println("=== Go Skynet Benchmark ===")
	fmt.Printf("Leaves: %d, concurrency units: %d (sequential below %d)\n\n",
		numLeaves, totalActors, seqThreshold)

	root := make(chan int64, 1)
	start := time.Now()
	go skynetNode(root, 0, numLeaves)
	sum := <-root
	elapsed := time.Since(start)

	elapsedNs := elapsed.Nanoseconds()
	elapsedUs := elapsedNs / 1000

	fmt.Printf("Sum: %d\n", sum)
	if elapsedUs > 0 {
		nsPerMsg := elapsedNs / rateBase
		throughputM := rateBase / elapsedUs
		leftover := rateBase - (throughputM * elapsedUs)
		throughputFrac := (leftover * 100) / elapsedUs
		fmt.Printf("ns/msg:         %d\n", nsPerMsg)
		fmt.Printf("Throughput:     %d.", throughputM)
		if throughputFrac < 10 {
			fmt.Printf("0")
		}
		fmt.Printf("%d M msg/sec\n", throughputFrac)
	}
}

File: benchmarks/cross-language/rust/src/skynet.rs
// Rust Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Uses std::thread while the subtree is larger than SEQ_THRESHOLD.
// Spawning 1M OS threads is not feasible; this limits concurrent threads to ~1000.

use std::sync::mpsc::{channel, Sender};
use std::thread;
use std::time::Instant;

fn get_leaves() -> i64 {
    if let Ok(s) = std::env::var("SKYNET_LEAVES") {
        if let Ok(n) = s.parse() {
            return n;
        }
    }
    if let Ok(s) = std::env::var("BENCHMARK_MESSAGES") {
        if let Ok(n) = s.parse() {
            return n;
        }
    }
    1_000_000
}

// Same threshold in every language here; divisor is the unit count, not the
// tree's nodes. See FAIRNESS.md.
const SEQ_THRESHOLD: i64 = 1000;

fn skynet_seq(offset: i64, size: i64) -> i64 {
    let mut sum = 0i64;
    for i in 0..size {
        sum += offset + i;
    }
    sum
}

fn skynet(tx: Sender<i64>, offset: i64, size: i64, depth: usize) {
    if size <= SEQ_THRESHOLD {
        tx.send(skynet_seq(offset, size)).unwrap();
        return;
    }
    let child_size = size / 10;
    let (child_tx, child_rx) = channel::<i64>();
    for i in 0..10i64 {
        let child_offset = offset + i * child_size;
        let child_tx = child_tx.clone();
        thread::spawn(move || {
            skynet(child_tx, child_offset, child_size, depth + 1);
        });
    }
    let mut sum = 0i64;
    for _ in 0..10 {
        sum += child_rx.recv().unwrap();
    }
    tx.send(sum).unwrap();
}

fn main() {
    let num_leaves = get_leaves();

    // Units created; also the divisor. See FAIRNESS.md.
    let mut total_actors = 1i64;
    let mut n = num_leaves;
    while n > SEQ_THRESHOLD {
        total_actors += n / SEQ_THRESHOLD;
        n /= 10;
    }

    println!("=== Rust Skynet Benchmark ===");
    println!("Leaves: {}, concurrency units: {} (sequential below {})\n", num_leaves, total_actors, SEQ_THRESHOLD);

    let (tx, rx) = channel::<i64>();
    let start = Instant::now();
    thread::spawn(move || skynet(tx, 0, num_leaves, 0));
    let sum = rx.recv().unwrap();
    let elapsed = start.elapsed();

    let elapsed_ns = elapsed.as_nanos() as i64;
    let elapsed_us = elapsed_ns / 1000;

    println!("Sum: {}", sum);
    if elapsed_us > 0 {
        let ns_per_msg = elapsed_ns / total_actors;
        let throughput_m = total_actors / elapsed_us;
        let leftover = total_actors - (throughput_m * elapsed_us);
        let throughput_frac = (leftover * 100) / elapsed_us;
        print!("ns/msg:         {}\n", ns_per_msg);
        print!("Throughput:     {}.", throughput_m);
        if throughput_frac < 10 {
            print!("0");
        }
        println!("{} M msg/sec", throughput_frac);
    }
}
